src/GraphQL/Resolver.hs

```diff
@@ -326,3 +330,98 @@ instance forall typeName interfaces fields m.
+
+instance forall m union os n i f uName objs.
+  ( Monad m
+  , KnownSymbol n
+  , TypeIndex m (API.Object n i f) union ~ Handler m (API.Object n i f)
+  , RunFields m (RunFieldsType m f)
+  , RunUnion m union os
+  ) => RunUnion m union ((API.Object n i f):os) where
+  runUnion duv@(DynamicUnionValue label _) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType name') [] selection)) =
+    if label == AST.getNameText name'
+    then case extractUnionValue @(API.Object n i f) @union @m duv of
+           Nothing -> runUnion @m @union @os duv fragment
+           Just handler -> buildResolver @m @(API.Object n i f) handler selection
+    else runUnion @m @union @os duv fragment -- TODO is this the error condition?
+
+instance forall m union. RunUnion m union '[] where
+  runUnion _ _ = notImplemented -- TODO error case
+
+-- TODO empty list of RunUnion
+
+-- TODO: list version (I think)
+instance forall m unionName objects.
+  ( Monad m
+  , KnownSymbol unionName
+  , RunUnion m (API.Union unionName objects) objects
+  ) => HasGraph m (API.Union unionName objects) where
+  type Handler m (API.Union unionName objects) = DynamicUnionValue (API.Union unionName objects) m
+  buildResolver mHandler selectionSet = do
+    let duv@(DynamicUnionValue label _) = mHandler
+    -- we only need to look at the fragment that matches:
+    case find (matchFragmentName label) selectionSet of
+      Nothing -> pure (Result [UnionTypeNotFound] GValue.ValueNull) -- TODO more error detail
+      Just fragment -> do
+        -- loop through union handlers and call right one when type matches.
+        runUnion @m @(API.Union unionName objects) @objects duv fragment
+
+      where
+        matchFragmentName label' (AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType name') [] selection)) =
+          label' == AST.getNameText name'
+
+
+symbolText :: forall ks. KnownSymbol ks => Text
+symbolText = toS (symbolVal @ks Proxy)
+
+-- | Build a tagged handler for an Object. This allows us to create a
+-- list of DynamicUnionValue that have the same type, but are backed
+-- by different handlers. We can then extract the correct handler
+-- based on a name (the name comes from an inline fragment).
+--
+-- Use e.g. like "unionValue @O1 @U1" - the @U1 is not neccessary when
+-- GHC can infer it.
+unionValue ::
+  forall (o :: Type) (union :: Type) m (name :: Symbol) i f.
+  (Monad m, API.Object name i f ~ o, KnownSymbol name)
+  => TypeIndex m o union -> DynamicUnionValue union m
+unionValue x =
+  DynamicUnionValue (symbolText @name) (unsafeCoerce x)
+
+extractUnionValue ::
+  forall (o :: Type) (union :: Type) m (name :: Symbol) i f.
+  (Monad m, API.Object name i f ~ o, KnownSymbol name) -- TODO probably don't need all constraints here
+  => DynamicUnionValue union m -> Maybe (TypeIndex m o union)
+extractUnionValue (DynamicUnionValue uName uValue) =
+  if uName == symbolText @name
+  then Just (unsafeCoerce uValue)
```

Comment:
sorry yea, it's coming - I wasn't going to write comments until I knew the scheme worked.